relay-pattern/src/wildmatch.rs

```diff
@@ -109,6 +109,27 @@ where
                     // No match, allow for backtracking.
                     false
                 }
+                Token::Optional(inner) => {
+                    let empty = &[Token::Literal(Literal(String::from("")))][..];
+                    let matches = match &**inner {
+                        Token::Alternates(alternates) => iter::once(empty)
+                            .chain(alternates.iter().map(|alt| alt.as_slice()))
+                            .any(|alternate| {
+                                let tokens = tokens.with_alternate(t_next, alternate);
+                                is_match_impl::<_, M>(h_current, &tokens)
+                            }),
+                        _ => [empty, slice::from_ref(&**inner)].iter().any(|alternate| {
+                            let tokens = tokens.with_alternate(t_next, alternate);
+                            is_match_impl::<_, M>(h_current, &tokens)
+                        }),
+                    };
+                    // The brace match already matches to the end, if it is successful we can end right here.
+                    if matches {
+                        return true;
+                    }
+                    // No match, allow for backtracking.
+                    false
+                }
```

Comment:
```suggestion
                Token::Optional(optional) => {
                    let optional = tokens.with_alternate(t_next, std::slice::from_ref(optional));
                    if is_match_impl::<_, M>(h_current, &optional) {
                        // There is a match with the optional token, we're done.
                        return true;
                    }
                    // Continue on without the optional token.
                    true
                }
```

Reply:
Yes, that is _much_ better. Part of why I made it so complicated is not seeing the forest for the trees, but another part is also the fact that I thought we wouldn't be able to just call `with_alternate` with the optional token if it contains an `Alternate` token, and to be honest I still don't quite understand why that works.